### PR TITLE
Datasets that extract fixed-duration windows

### PR DESCRIPTION
**Pull request by matt-har-vey**
This commit adds standalone functions that extract random or sliding
context windows of a common duration from features parsed from
tf.train.Examples produced by examplegen.

It also includes an initialization method, new_window_dataset, that
provides a similified API for constructing context-window-level traning
and validation sets.

Finally, it adds unit tests for all methods in the dataset module.

Change-Id: I5c6abd668ded92d8004d9f0dd332d394d33bf9d5

### CODE REVIEW

File: multispecies_whale_detection/dataset.py [hunk 1]
```diff
@@ -70,3 +92,257 @@ def new(tfrecord_filepattern: str):
...
+
+  Returns:
+    Audio tensor of shape [sample_size, audio_duration_samples] and labels
+    tensor of float32 class labels of shape [sample_size, len(class_names)].
+  """
+  num_hops = 1 + tf.cast(
+      tf.math.floor(
+          (_audio_duration(features) - context_duration_seconds) / hop_seconds),
+      tf.int32)
+  extracted_windows = tf.TensorArray(tf.float32, size=num_hops)
```
Comment (sdenton4) [inline multispecies_whale_detection/dataset.py]: (neat! I don't think I've seen these before.)

File: multispecies_whale_detection/dataset.py [hunk 2]
```diff
@@ -70,3 +92,257 @@ def new(tfrecord_filepattern: str):
...
+    tfrecord_filepattern: str,
+    context_duration_seconds: float,
+    class_names: Sequence[str],
+    sample_size: Optional[int] = None,
+    hop_seconds: Optional[float] = None,
+    min_overlap=DEFAULT_MIN_OVERLAP,
+) -> tf.data.Dataset:
+  """Parallel tf.data pipeline to extract windows from TFRecords.
+
+  This is intended for initializing a training set when sample_size is not None
```
Comment (sdenton4) [inline multispecies_whale_detection/dataset.py:308]: How would you feel about making the intention explicit with a train/test mode flag (or is_train boolean, which seems more keras-y)? Then we can have arguments train_sample_size and eval_hop_seconds.

Rationale: It's nice to be able to configure options without worrying about mutually exclusive arguments, like new_window_dataset(mode=TRAIN, **config_dict).

File: multispecies_whale_detection/dataset.py [hunk 3]
```diff
@@ -70,3 +92,257 @@ def new(tfrecord_filepattern: str):
...
+    window_fn = functools.partial(random_windows, **window_fn_args)
+  else:
+    window_fn_args.update(hop_seconds=hop_seconds)
+    window_fn = functools.partial(sliding_windows, **window_fn_args)
+
+  def shard_fn(filename):
+    dataset = tf.data.TFRecordDataset(filename)
+    dataset = dataset.map(parse_fn)
+    dataset = dataset.map(window_fn)
+    dataset = dataset.unbatch()
```
Comment (sdenton4) [inline multispecies_whale_detection/dataset.py:351]: You may want a prefetch here to allow buffering the disk IO and parsing (similar to using readahead).

**Review by sdenton4 — APPROVED** [review]

File: multispecies_whale_detection/dataset.py (hunk 1 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/dataset.py]: Thanks.

It was mostly new to me as well, though I think I've seen it once before around a while loop in the PCEN IIR filter.

My first attempt here used ```for _ in range(num_hops)```, and it caused an error of the form "this is not in the same graph as that." I think this is because num_hops is not statically known because the clip duration isn't. There's an option to add that constraint and make users pass a static clip duration down from the main method (flags), but I don't think this is so much less efficient as to warrant that tiny extra burden.

(I haven't tested this on TPU and would need to be sure it works.)

File: multispecies_whale_detection/dataset.py (hunk 3 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/dataset.py:351]: (To avoid doubt, what follows is my fastidiousness showing.)

I might want it, but I'm not convinced it's necessary (in the mathematical sense of that word).

An expectation I maybe should have made explicit in the docstring is that a caller who is using this for training input will .shuffle().batch().prefetch(), where probably the shuffle is somewhere in (batch size, 2 * batch size) and the prefetch is 1 batch. In many cases I have found that enough to get model-bound.

TFRecordDataset(buffer_size=) is more targeted at decoupling from file reads ("disk") than is prefetch. I like defaults until they prove insufficient.

It's probably OK for record slicing, proto parsing, and Tensor filling to block on each other - because they aren't blocking any of the other parallel calls in the interleave. One big enough buffer (shuffle in train or prefetch in eval) from the caller should be enough to even it all out.

(Appendix: Is .batch().map(parse_example).unbatch() lots more efficient that .map(parse_single_example)? I don't know, but presumably there's a reason for batch and single versions of the same thing to exist. It doesn't seem intuitive that proto parsing would parallelize well.)

File: multispecies_whale_detection/dataset.py (hunk 2 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/dataset.py:308]: Point taken on the bad aesthetics of mutually exclusive arguments.

However, I don't find it much prettier to add a third one to disambiguate which of the other two is ignored. I also don't love TRAIN/train_/eval_ naming for assuming the caller's use case, but I think this wasn't your main point.

What I really want is perfectly valid Rust

```
enum Windowing {
  RANDOM(u8),
  SLIDING(f32),
}
```

or C++ std::variant. I don't know whether such a thing exists in Python, do you?

Comment (matt-har-vey) [reply multispecies_whale_detection/dataset.py:308]: I just found [pattern matching](https://www.python.org/dev/peps/pep-0636/#adding-a-ui-matching-objects) is [supposed to come to Python 3.10](https://mail.python.org/archives/list/python-committers@python.org/message/SQC2FTLFV5A7DV7RCEAR2I2IKJKGK7W3/). For now, I'll [use isinstance](http://blog.ezyang.com/2020/10/idiomatic-algebraic-data-types-in-python-with-dataclasses-and-union/).

Comment (matt-har-vey) [reply multispecies_whale_detection/dataset.py:308]: I guess object-oriented (strategy pattern) is the Pythonic way to do this, unless we really needed "windowing" to be a data object, but there's no real need for that yet.

**Review by matt-har-vey — COMMENTED** [review]
PTAL because the changes are substantial.

Sorry for getting carried away, but I find it much cleaner now.
